Feature/timi/search guide (#605)
MapsIndoors/docs: src/ios/v3/guides/search/create-a-search-experience-with-mapsindoors.md

---
title: Create a search experience with MapsIndoors
eleventyNavigation:
  key: ios-v3-guides-search-experience
  parent: ios-v3-guides-search
  title: Create a search experience with MapsIndoors
  order: 510
---

> This is an example of creating a simple search experience using MapsIndoors. We will create a map with a search button that leads to another view controller that handles the search and selection. On selection of a location, we go back to the map and shows the selected location on the map.

Reviewer: ```suggestion
This is an example of creating a simple search experience using MapsIndoors. We will create a map with a search button which leads to another view controller that handles the search and selection. Select a Location togo back to the map and show the selected Location on the map.
```

> We will start by creating a simple search controller that handles search and selection of MapsIndoors locations

Reviewer: ```suggestion
We will start by creating a simple search controller that handles search and selection of MapsIndoors Locations.
```

> Declare a protocol for our location selection with a `didSelectLocation` method

Reviewer: ```suggestion
Declare a protocol for our Location selection with a `didSelectLocation` method
```

```swift
protocol MySearchControllerDelegate {
    func didSelectLocation(location:MPLocation)
}
```

> Define `MySearchController`. In this tutorial our search controller is a `UIViewController` that implements the protocols `UISearchBarDelegate`, `UITableViewDelegate` and `UITableViewDataSource`

Reviewer: ```suggestion
Define `MySearchController`. In this tutorial our search controller is a `UIViewController` that implements the protocols `UISearchBarDelegate`, `UITableViewDelegate` and `UITableViewDataSource`.
```

```swift
class MySearchController: UIViewController, UISearchBarDelegate, UITableViewDelegate, UITableViewDataSource {
```

Setup member variables for `MySearchController`:

* An instance of type `MPLocationService`
* An instance of type `MPQuery`
* An array of `MPLocation` to hold your list of results
* Your delegate object
* A search bar view
* A table view

```swift
let locationService = MPLocationService.sharedInstance()
let query = MPQuery.init()
var locations:[MPLocation] = []
var delegate:MySearchControllerDelegate? = nil
let tableView = UITableView.init()
let searchBar = UISearchBar.init()
```

> In `viewDidLoad`, wire up your view controller to the tableview and search bar.

Reviewer: ```suggestion
In `viewDidLoad`, wire up your view controller to the table view and search bar.
```

        filter.take = 10
        locationService.getLocationsUsing(query, filter: filter) { (locations, error) in
            if error == nil {
                self.locations = locations!
                self.tableView.reloadData()
            }
        }
    } else {
        self.locations = []
        self.tableView.reloadData()
    }
}
```

Implement the `searchBarCancelButtonClicked` method, with dismissal of the view controller.

```swift
func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
    self.dismiss(animated: true, completion: nil)
}
```

Implement the `tableView:cellForRowAt` method. Set the `cell.textLabel.text` to reflect the *name* of the location of same index.

```swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "reuseIdentifier", for: indexPath)
    cell.textLabel?.text = locations[indexPath.row].name
    let defaultValue = ""
    cell.textLabel?.text?.append(", \(locations[indexPath.row].roomId ?? defaultValue), \(locations[indexPath.row].building ?? defaultValue), \(locations[indexPath.row].venue ?? defaultValue)")
    return cell
}
```

> Implement the `tableView:didSelectRowAt` method. In this example we just call the delegate method and dismiss the view controller.

Reviewer: ```suggestion
Implement the `tableView:didSelectRowAt` method. In this example we call the delegate method and dismiss the view controller.
```
